pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -99,31 +94,39 @@ type stateData struct {
 	unavailableClaims sets.Set[int]
 
 	informationsForClaim []informationForClaim
+
+	// nodeAllocations caches the result of Filter for the nodes.
+	nodeAllocations map[string][]*resourceapi.AllocationResult
+
+	// filterErr is set by Filter if a CEL runtime error is encountered
+	// for at least one node. Such an error indicates that something is
+	// wrong with the input (class or claim) and thus scheduling should not
+	// proceed, even if the error did not occur on all nodes.
```

Comment:
I would treat it as a fatal error and check in PreFilter instead of Filter. If somehow we _have_ to check in Filter, I guess there is no hook in Filer to "cancel" the parent call immediately. Also, a kind of this error would abort the whole scheduling lifecycle right?

Reply:
This is a CEL runtime error which cannot be detected in `PreFilter`. As you said, there is no mechanism at the moment for a `Filter` implementation to "abort" the entire scheduling.

Aborting is the intent here. Something in the CEL expression provided by the user is flawed because runtime errors should not occur. They can occur with some nodes, but not others, for example because the CEL expression gets evaluated against a device which only exists on some of the nodes.

There are two options:
- ignore nodes where the runtime error occurs, picking some other node if there is one
- abort (current behavior)

I think it's better to abort. Otherwise the user might have made some subtle mistake that prevents achieving what they want, without that being obvious because the runtime error will not be visible as long as there are other nodes and scheduling will appear to be working - until it doesn't.

We have [an issue](https://github.com/kubernetes/kubernetes/issues/125826) open about exposing problems with CEL better, perhaps when we get to that (not in 1.31!) then we can be more lenient and allow scheduling to "limp along" despite such a flawed CEL expression.

Reply:
I think we can abort is with an error returned, and it's a true error actually.
https://github.com/kubernetes/kubernetes/blob/bc3c07091b3501ac52c7c322f246df5e36815c20/pkg/scheduler/schedule_one.go#L622-L624

Reply:
So all it takes is to return a `framework.Error` instead of a `framework.Unschedulable`? I was following this documentation:
https://github.com/kubernetes/kubernetes/blob/bc3c07091b3501ac52c7c322f246df5e36815c20/pkg/scheduler/framework/interface.go#L425-L428

It clearly says that `framework.Error` isn't special and only excludes the node - probably worth fixing separately.

What you pointed out makes it simpler than I had thought... let me try it. Indeed, it works:
```
  Warning  FailedScheduling             4s (x11 over 80s)  default-scheduler  running "DynamicResources" filter plugin: claim dra-7783/tester-1-my-inline-claim-p8s2x: selector #0: CEL runtime error: no such key: exists
```

Here's the patch for the plugin that I'll squash into the commit:
```patch
diff --git a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
index e0d99c975ba..ac537cbaa13 100644
--- a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
+++ b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
@@ -97,15 +97,6 @@ type stateData struct {
 
 	// nodeAllocations caches the result of Filter for the nodes.
 	nodeAllocations map[string][]*resourceapi.AllocationResult
-
-	// filterErr is set by Filter if a CEL runtime error is encountered
-	// for at least one node. Such an error indicates that something is
-	// wrong with the input (class or claim) and thus scheduling should not
-	// proceed, even if the error did not occur on all nodes.
-	//
-	// This error gets returned by PostFilter as-is and thus should
-	// contain all relevant details.
-	filterErr error
 }
 
 func (d *stateData) Clone() framework.StateData {
@@ -940,10 +931,7 @@ func (pl *dynamicResources) Filter(ctx context.Context, cs *framework.CycleState
 			//
 			// Also record the error in the state to ensure that PreScore sees it and
 			// can prevent scheduling from continuing.
-			state.mutex.Lock()
-			defer state.mutex.Unlock()
-			state.filterErr = fmt.Errorf("filter node %s: %w", node.Name, err)
-			return statusUnschedulable(logger, err.Error(), "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaims", klog.KObjSlice(state.allocator.ClaimsToAllocate()))
+			return statusError(logger, err, "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaims", klog.KObjSlice(state.allocator.ClaimsToAllocate()))
 		}
 		// Check for exact length just to be sure. In practice this is all-or-nothing.
 		if len(a) != len(state.allocator.ClaimsToAllocate()) {
@@ -1059,9 +1047,6 @@ func (pl *dynamicResources) PreScore(ctx context.Context, cs *framework.CycleSta
 	}
 
 	logger := klog.FromContext(ctx)
-	if state.filterErr != nil {
-		return statusUnschedulable(logger, state.filterErr.Error())
-	}
 	pending := false
 	for index, claim := range state.claims {
 		if claim.Status.Allocation == nil &&
@@ -1142,10 +1127,6 @@ func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleStat
 	}
 
 	logger := klog.FromContext(ctx)
-	// PreScore might not have been called when there was only one node.
-	if state.filterErr != nil {
-		return statusUnschedulable(logger, state.filterErr.Error())
-	}
 
 	numDelayedAllocationPending := 0
 	numClaimsWithStatusInfo := 0
```

Reply:
There's one small drawback of this approach: it puts the pod into the backoff queue, where it will be retried. With the approach that I had before, the retries get avoided until the claim or class causing the problem get updated.

But it's simpler and deals with something that shouldn't occur in practice, so let's go with "simple".